### PR TITLE
allow gather_for_metrics to be more flexible

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do ? 
This Pr makes `gather_for_metrics` more flexible when it comes to tensors. We introduce `use_gather_object` flag that enable users to gather object even if the input data is composed of tensors only. This is useful in the case the tensors have different sizes and we don't wanna end up padding all tensors to the same shape/size as it might be a bit complicated afterwards. Note that this is not very efficient for tensors on GPU. Also, we make sure that we do not adjust the sample when we are not in distributed setup. 

cc [qubvel](https://github.com/qubvel) could you try your use case by setting `use_gather_object = True` ? 

Fixes https://github.com/huggingface/transformers/issues/30357

## Example where we should use `gather_object` instead of `gather` in `gather_for_metrics`

```python
import torch
import random
from torch.utils.data import DataLoader, Dataset
from accelerate import Accelerator


class Dataset(Dataset):
    def __init__(self, size):
        self.size = size

    def __len__(self):
        return self.size

    def __getitem__(self, idx):
        return {"boxes": torch.rand(idx, 4)}


dataloader = DataLoader(Dataset(5), batch_size=4, collate_fn=lambda x: x)
accelerator = Accelerator()
dataloader = accelerator.prepare(dataloader)

for i, batch in enumerate(dataloader):
    #skip until the last batch
    if i != len(dataloader) - 1:
        continue
    
    print("Batch before gather:", [x["boxes"].shape for x in batch])
    gathered_batch = accelerator.gather_for_metrics(batch, use_gather_object=True)
    print("Batch after gather: ", [x["boxes"].shape for x in gathered_batch])

```

With two processes, we get: 
```
Batch before gather: [torch.Size([0, 4]), torch.Size([1, 4]), torch.Size([2, 4]), torch.Size([3, 4])]
Batch before gather: [torch.Size([4, 4]), torch.Size([0, 4]), torch.Size([1, 4]), torch.Size([2, 4])]
Batch after gather:  [torch.Size([0, 4]), torch.Size([1, 4]), torch.Size([2, 4]), torch.Size([3, 4]), torch.Size([4, 4])]
Batch after gather:  [torch.Size([0, 4]), torch.Size([1, 4]), torch.Size([2, 4]), torch.Size([3, 4]), torch.Size([4, 4])]
```

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -2243,14 +2243,18 @@ def gather(self, tensor):
         """
         return gather(tensor)
 
-    def gather_for_metrics(self, input_data):
+    def gather_for_metrics(self, input_data, use_gather_oject=False):
```
Comment (muellerzr) [inline src/accelerate/accelerator.py:2246]:
```suggestion
    def gather_for_metrics(self, input_data, gather_non_tensors=False):
```

Perhaps using `gather_non_tensors` instead :)

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -2289,12 +2295,16 @@ def gather_for_metrics(self, input_data):
                         "The used dataset had no length, returning gathered tensors. You should drop the remainder yourself."
                     )
                     return data
-                elif self.gradient_state.remainder > 0:
+                elif self.gradient_state.remainder > 0 and self.distributed_type != DistributedType.NO:
```
Comment (muellerzr) [inline src/accelerate/accelerator.py:2298]: Nice I think this winds up solving another issue I saw somewhere about samples being dropped

**Review by muellerzr — APPROVED** [review]
As long as tests pass, LG2m!

File: src/accelerate/accelerator.py (hunk 1 above)
Comment (SunMarc) [reply src/accelerate/accelerator.py]: I think this can be confusing to ask the user to set `gather_non_tensors=True` with tensors.

Comment (muellerzr) [reply src/accelerate/accelerator.py]: Fair point :)

File: src/accelerate/accelerator.py [hunk 3]
```diff
@@ -2275,7 +2280,9 @@ def gather_for_metrics(self, input_data):
         except TypeError:
             all_tensors = False
 
-        if not all_tensors:
+        use_gather_oject = use_gather_oject or not all_tensors
```
Comment (qubvel) [inline src/accelerate/accelerator.py:2283]: I guess some typo: use_gather_oject -> use_gather_o**b**ject

Comment (SunMarc) [reply src/accelerate/accelerator.py]: Oh indeed thx !

File: src/accelerate/accelerator.py [hunk 4]
```diff
@@ -2243,14 +2243,19 @@ def gather(self, tensor):
...
+    def gather_for_metrics(self, input_data, use_gather_object=False):
         """
         Gathers `input_data` and potentially drops duplicates in the last batch if on a distributed system. Should be
         used for gathering the inputs and targets for metric calculation.
 
         Args:
             input (`torch.Tensor`, `object`, a nested tuple/list/dictionary of `torch.Tensor`, or a nested tuple/list/dictionary of `object`):
                 The tensors or objects for calculating metrics across all processes
+            use_gather_object(`bool`):
+                Whether to use or not gather_object instead of gather. This flag can be useful for gathering tensors
```
Comment (muellerzr) [inline src/accelerate/accelerator.py:2255]:
```suggestion
                Whether to forcibly use gather_object instead of gather (which is already done if all objects passed do not contain tensors). This flag can be useful for gathering tensors
```

Comment (qubvel) [reply src/accelerate/accelerator.py]: Was not very clear to me unless I read `gather_object` description, maybe worth including it also and/or adding an example?
`Recursively gather object in a nested list/tuple/dictionary of objects from all devices.`

**Review by muellerzr — APPROVED** [review]

**Comment by hujh1994** [comment]
> try your use case by setting `use_gather_object = True`

I tried his use case by setting use_gather_object = True. It works！
